### ISSUE TITLE
Default concurrent mode misses TS2307 that `--singleThreaded` (and TS 6.0) report, for import/export declarations inside non-scope blocks

### ISSUE THREAD

**Issue by mohsen1** [microsoft/typescript-go#4834]
## Steps to reproduce

`bareblock.ts`:
```ts
{
    export { a } from "m";
}
```

```
tsc --noEmit --pretty false bareblock.ts
tsc --noEmit --pretty false --singleThreaded bareblock.ts
```

Reproduces on `typescript@7.0.2` and `7.1.0-dev.20260805.1`. Both modes are 100% deterministic — the diagnostic *set* differs, not the ordering.

## Behavior with `typescript@6.0`

```
bareblock.ts(2,5): error TS1233: An export declaration can only be used at the top level of a namespace or module.
bareblock.ts(2,23): error TS2307: Cannot find module 'm' or its corresponding type declarations.
```

## Behavior with `tsgo`

Default (concurrent) mode drops the TS2307:

```
bareblock.ts(2,5): error TS1233: An export declaration can only be used at the top level of a namespace or module.
```

With `--singleThreaded`, output matches `typescript@6.0` exactly:

```
bareblock.ts(2,5): error TS1233: An export declaration can only be used at the top level of a namespace or module.
bareblock.ts(2,23): error TS2307: Cannot find module 'm' or its corresponding type declarations.
```

Since the single-threaded output matches Strada, the default mode looks like a false negative: users silently lose "Cannot find module" for this shape.

## Scope

Divergence occurs for import/export declarations (`export { a } from "m"`, `export * from "m"`, `import { a } from "m"`, `import x = require("m")`, …) inside any container that does not open a declaration scope: bare block, `if`/`for`/`while`/`try` bodies, labeled block, `switch` clause. Containers that open a scope (function/method body, class static block, namespace body, ambient module) agree in both modes, as does side-effect `import "m";`.

**Comment by itibbers** [comment]
## Summary

TypeScript 7.0.2 produces **non-deterministic** type-checking output across repeated runs with identical source and config. The instability only appears in the default multi-threaded mode **and only when `allowJs: true`** is set. Disabling `allowJs` or running with `--singleThreaded` makes the output fully deterministic, indicating a race condition in the multi-threaded `allowJs` type-inference path.

## My test

Run `tsc --pretty false` repeatedly in the same project without touching any file:

| Mode | Run 1 | Run 2 | Run 3 | Run 4 | Run 5 | Run 6 | Run 7 | Run 8 | Stable? |
|------|-------|-------|-------|-------|-------|-------|-------|-------|---------|
| Multi-threaded (default), `allowJs: true` | 10091 | 10122 | 10138 | 10195 | 10199 | 10200 | 10202 | 10448 | ❌ Fluctuates by hundreds |
| Single-threaded (`--singleThreaded`), `allowJs: true` | 10531 | 10531 | 10531 | — | — | — | — | — | ✅ Identical |
| Multi-threaded (default), `allowJs: false` | 12817 | 12817 | 12817 | 12817 | 12817 | — | — | — | ✅ Identical |
| Single-threaded (`--singleThreaded`), `allowJs: false` | 12812 | 12812 | 12812 | — | — | — | — | — | ✅ Identical |

(Numbers are total `error TS` counts.)
